lms/lmsdb/models.py

```diff
@@ -328,15 +328,40 @@ def is_checked(self):
     def create_hash(content: Union[str, bytes], *args, **kwargs) -> str:
         return hashing.by_content(content, *args, **kwargs)
 
+    @classmethod
+    def _is_previous_submission_exists(
+        cls, user: User, exercise: Exercise,
+    ) -> bool:
+        return (
+            cls
+            .select()
+            .where(
+                cls.solver == user,
+                cls.exercise == exercise,
+            )
+        ).exists()
+
     @classmethod
     def is_duplicate(
-            cls, content: Union[str, bytes], user: User, *,
-            already_hashed: bool = False,
+            cls, content: Union[str, bytes], user: User, exercise: Exercise,
+            *, already_hashed: bool = False,
     ) -> bool:
+
+        if not cls._is_previous_submission_exists(user, exercise):
```

Comment:
Is this necessary?

(I try to considering the fact that 2 queries are probably much more expensive than a single one)

Reply:
It isn't, since this case is covered by the query in the is_duplicate function. The upside of it is that in cases the user hasn't submitted any solution before, we'll only run the rather simple (I think) query in the is_previous_submission_exists function.

Would you prefer I'll drop this part?

Reply:
This is a legitimate consideration which should be profiled.

The performance here depends on the user's usage style. It could be the case that in one instance of LMS users are mostly reuploading solutions for one exercise, and in another instance the users usually send only one solution to each exercise.

I think that, for simplification, we should drop this query :)